### PR TITLE
[GlobalOptimization] Support SetEncoding on batch matmul cases with producer CastOpInterface ops

### PR DESCRIPTION
**Pull request by Max191**
This PR completes the SetEncoding plumbing discussed in https://github.com/openxla/iree/issues/15214

### CODE REVIEW

File: compiler/src/iree/compiler/GlobalOptimization/ExpandVectors.cpp [hunk 1]
```diff
@@ -11,13 +11,45 @@
...
 #include "mlir/Dialect/Tensor/IR/Tensor.h"
 #include "mlir/Transforms/GreedyPatternRewriteDriver.h"
 
 namespace mlir {
 namespace iree_compiler {
 namespace GlobalOptimization {
 
+// If the producer is a CastOpInterface, or a linalg::GenericOp that performs
+// only a CastOpInterface on its input, return the CastOpInterface op
+static Operation *getDefiningCastOp(Value input) {
```
Comment (Max191) [inline compiler/src/iree/compiler/GlobalOptimization/ExpandVectors.cpp:24]: Where should a util function like this live? It is somewhat specific to this matching, but it is needed in both ExpandVectors.cpp and SetEncoding.cpp now. Should we make a GlobalOptimization utils file?

Comment (hanhanW) [reply compiler/src/iree/compiler/GlobalOptimization/ExpandVectors.cpp:24]: Yes, please create GlobalOptmization/Utils.[h|cpp] files and move the functions to there.

**Comment by Max191** [comment]
We can split this into 2 PRs if that makes more sense. I put the changes together for now, since they are related.

**Comment by hanhanW** [comment]
> We can split this into 2 PRs if that makes more sense. I put the changes together for now, since they are related.

Given that you already have two separate commits, can you split it into 2 PRs?

**Comment by Max191** [comment]
This PR has been split into https://github.com/openxla/iree/pull/15371 and https://github.com/openxla/iree/pull/15372 now

**Pull request closed without merging**